### ISSUE TITLE
Typesafe babel plugin development with TypeScript

### ISSUE THREAD

**Issue by hbroer** [babel/babel#10637]
## Feature Request

**Is your feature request related to a problem? Please describe.**
I miss some types for plugin development in typescript which I need. I asked for some help at the slack channel but got no answer.

1. I can't find the type for PluginPass (https://github.com/babel/babel/blob/master/packages/babel-core/src/transformation/plugin-pass.js). It should be exported to make it usable like in this example:

```typescript
import * as types from "@babel/types";
import {ImportDeclaration} from "@babel/types";
import {NodePath, PluginPass} from "@babel/traverse";
// or import {NodePath, PluginPass} from "@babel/core";

export default function({ types: t }: {types: typeof types}) {
    return {
        name: "example plugin",
        visitor: {
            ImportDeclaration(path: NodePath<ImportDeclaration>, pass: PluginPass) {}
        }
    };
}
```

2. The plugin function has no return type but it should to display errors and to make code completion possible, like this:

```typescript
import {PluginDefinition} from "@babel/core";

export default function(): PluginDefinition {
    return {}; // this should be a TS error because `name` is missing
}
```

3. `{types: typeof types}` is strange also would be this: `typeof babel`. This is a workaround which describes the plugin default function parameter. This should also a dedicated type:

```typescript
import {PluginDefinition, Babel} from "@babel/core";

export default function({ types: t }: Babel): PluginDefinition {}
```

**Describe the solution you'd like**
Export/add the types which are missing.

**Describe alternatives you've considered**
None. Declare everything as `any` is not a solution :-)

**Teachability, Documentation, Adoption, Migration Strategy**
The documentation/handbook should contain a example how to write plugins in TypeScript.

**Comment by hbroer** [comment]
Just noticed that only 50% of the available Types come from @babel, the other 50% come from @types. :/

Also noticed one of Types I missed I just oversee: `PluginObj` is the type of the functionr eturn value. But still missing the `PluginPass` definition.

And now I don't know who to ask, or where to add, the missing type...

**Comment by eric-burel** [comment]
Crazy that there is not more docs about it. If only for auto-completion TS would help greatly. I've looked briefly at existing Babel plugin boilerplate but none is using TS. 

Are you aware of some plugins already using TypeScript?

**Comment by hbroer** [comment]
> Are you aware of some plugins already using TypeScript?

yea my tiny, still not finished lib ;)

https://github.com/owja/babel-css-inline-plugin

**Comment by hantatsang** [comment]
Would be great if babel's official Typescript definitions are complete, it's very confusing to use some types from `@babel` and some types from DefinitelyTyped (`@types`)

**Comment by danvoyce** [comment]
It does amaze me how such a popular library can be so far behind in the TS world.

**Comment by JLHwung** [comment]
Please be constructive :)

We are slowly migrating our codebase to TypeScript so hopefully we can provide better type annotations. You can track the progress in https://github.com/babel/babel/pulls?q=is%3Aopen+is%3Apr+label%3A%22Flow+-%3E+TS%22

Review and feedbacks to these PRs are always welcome.

**Comment by danvoyce** [comment]
Yeah apologies, I should have elaborated to avoid that sounding disrespectful or maybe I should have kept my thoughts to myself :)

I guess what I'm surprised about is that so many of us use this library on a daily basis, why can not more of us (myself included!) help improve it. 

I'm still getting up to speed with Typescript, but will take a look at the PR for sure. And if there's anything else I/we can do to help please shout!

**Comment by nbouvrette** [comment]
Until the TypeScript version is available there are still ways to make the code more readable, such as:

```ts
import type * as BabelCoreNamespace from '@babel/core';
import type * as BabelTypesNamespace from '@babel/types';
import type { PluginObj, PluginPass } from '@babel/core';

export type Babel = typeof BabelCoreNamespace;
export type BabelTypes = typeof BabelTypesNamespace;

export default function myPlugin(babel: Babel): PluginObj {
    return {
        name: "example",
        visitor: {
            ImportDeclaration(nodePath: NodePath<ImportDeclaration>, pluginPass: PluginPass) {
                // your code here.
            }
        }
    };
}
```

**Comment by ericbiewener** [comment]
My version of getting typesafety:

```tsx
import { NodePath, PluginObj } from "@babel/core";
import * as t from "@babel/types";

const plugin = (): PluginObj => ({
  name: "example",
  visitor: {
    // No need to type `nodePath` -- the `PluginObj` return type of the function provides typesafety on it already.
    ImportSpecifier(nodePath) {

    }
  }
})
```

The key here is that the `@babel/types` import must be the same version as the one used by `@babel/core` so that stuff like this will work:

```tsx
const isGqlTaggedTemplateExpression = (node: t.TaggedTemplateExpression) =>
  t.isIdentifier(node.tag) && node.tag.name === "gql";
```

If the versions are misaligned, trying to pass `path.node` to the above function from inside your plugin function will result in a TS error since `path.node` is getting its type from the version of `@babel/types` used by `@babel/core`.

**Comment by a7madgamal** [comment]
@JLHwung the link you posted seems done, does this mean this ticket is done?

**Comment by hungify** [comment]
I am feeling a bit confused.

> <img width="785" alt="CleanShot 2023-11-27 at 22 33 39@2x" src="https://github.com/babel/babel/assets/78008509/05f18868-dafc-4187-b635-e7861a8f9b83">

Note:

```typescript
import type { PluginObj } from "@babel/core"; // reference to @types/babel__core
import type * as BabelCoreNamespace from "@babel/core"; // reference to @types/babel__core
import type * as BabelTypesNamespace from "@babel/types"; // reference to @babel/types
```

**Comment by unphased** [comment]
I don't really need to get full type definitions for babel, but I would like to be able to import my typescript source code from my babel plugins as they become more sophisticated, and I'm not able to do that yet: 

     TypeError: [BABEL]: Unknown file extension ".ts"

**Comment by Den-dp** [comment]
One of the possible solutions, that requires `@babel/helper-plugin-utils`:

```ts
import { declare } from '@babel/helper-plugin-utils';

export default declare(({ types: t }) => {
  return {
    visitor: {
      StringLiteral(path, pass) {},
    },
  };
});
```
